### PR TITLE
Apply strain

### PR DESCRIPTION
**Pull request by samwaseda**
Apply a strain with a single number / 3-dim vector / epsilon matrix.

With `return_box` only the returned box will be strained

### CODE REVIEW

File: pyiron/atomistics/structure/atoms.py [hunk 1]
```diff
@@ -3106,6 +3106,31 @@ def set_constraint(self, constrain):
...
+                                              is applied in each direction. If a 3-dim vector is set, it
+                                              will be multiplied by a unit matrix.
+                return_box (bool): whether to return a box. If set to True, only the returned box will 
+                                   have the desired strain and the original box will stay unchanged.
+        """
+        epsilon = np.array([epsilon]).flatten()
+        if len(epsilon) == 3 or len(epsilon) == 1:
+            epsilon = epsilon*np.eye(3)
+        epsilon.reshape(3,3)
+        if epsilon.min()<-1.0:
```
Comment (liamhuber) [inline pyiron/atomistics/structure/atoms.py]: nit: `if epsilon.min() < -1.0`  whitespace

File: pyiron/atomistics/structure/atoms.py [hunk 2]
```diff
@@ -3106,6 +3106,31 @@ def set_constraint(self, constrain):
...
+            raise ValueError("Illegal strain value")
```
Comment (liamhuber) [inline pyiron/atomistics/structure/atoms.py]: nit: Can be more informative and let user know why it was illegal, i.e. too negative

**Comment by jan-janssen** [comment]
Looks good to me, maybe you can also update Murnaghan master to use this new method https://github.com/pyiron/pyiron/blob/master/pyiron/atomistics/master/murnaghan.py#L309

**Comment by samwaseda** [comment]
> Very nice. I would only add that I'd like to see this method called in `Atoms.__[r]mul__` and `Atoms__imul__` under the condition that the multiplying value is a 3x3 array or matrix. (And I guess we'd need to subtract `np.eye(3)` off of the multiplying value in that case).

That would mean that `cell` has to get `__mul__`, right? I guess that would slightly deviate from what is written in ASE, so I might want to avoid it (as I'm envisaging a convergence to ASE at some point....)

**Comment by liamhuber** [comment]
> That would mean that `cell` has to get `__mul__`, right?

Cell is type `numpy.ndarray` and already has the correspondning `__mul__` behaviour. If you mean we need to make a custom `Cell` class and modify it's `__mul__` behaviour, then no, I don't see why we'd need to do that. we can just do something like:

```python
Atoms.__mul__(self, other):
    # other stuff, then
    self.cell = np.matmul(self, other)
   # more other stuff
```

Right now as it is, people can directly modify a structure's cell, but it does exactly that. It modifies the cell and leaves the atoms alone. e.g.)

```python
struct = pr.create_ase_bulk('Al', cubic=True)
struct.cell *= 2
```

Just gives you a unit cell of atoms taking up one eights of a cubic box.

I can just as easily do `struct.positions *= np.pi` and get similar nonsense.

I'm only proposing that we give sophisticated behaviour to the `Atoms` class itself. If people want to mess around with the constituent arrays by themselves, then on their heads be it.

**Pull request merged**